### DISCUSSION TITLE
Designer questions & Workflows run at startup

### DISCUSSION THREAD

**Discussion by stevengwebsolutions** [elsa-workflows/elsa-core#1352] [Q&A]
Hello, I'm new to Elsa and i'm in the process is evaluating how it fits into my applications architecture.  I have a multitenancy .Net MVC app (using the AbpBoilerplate/NetZero framework) and we have a need to allow each tenant of our app to inject custom business logic, this includes validating objects based on their custom rules (field1 > 5 etc).  Providing the ability to allow our app tenants to use the Elsa Workflow designer to build validation workflows is perfect.  I will need to create workflow types (like ValidateDepartmentInput) and have each tenant create a version of that workflow type via the designer and then save that customized workflow definition.  Then at run time (when a user creates a new Department for example) I need to determine if the tenant has a workflow definition of type ValidateDepartmentInput, retrieve it from the definition store and then run it.  I'll have to figure out a naming convention to store all these similar (but different by tenant) workflows, in my code I'm planning on finding and invoking a workflow like this (unless there is a better way to find and run a WF from the definition store):

```
var workflows = await _workflowRegistry.ListActiveAsync();
var workflow = workflows.Where(w => w.Name == "ValidateDepartmentDemo" && w.IsPublished == true && w.IsLatest == true && w.TenantId == "99").Single();
var department = await _myAppService.GetAsync(new Abp.Application.Services.Dto.EntityDto { Id = 10 });
var startResult = await _startsWorkflow.StartWorkflowAsync(workflow, null, new WorkflowInput { Input = department });

```
I've got three issues that I need some help with:

1) Is it possible to use the workflow designer outside of the full workflow dashboard (on my 'manage department validation workflow' page I want to let the tenant launch the designer to specifically create a workflow of that type so my code can handle the tracking and persistence of that workflow type for the tenant.

2) Is it possible to point the designer at a faux server so that again I can encapsulate the designers interactions and add an additional management layer to workflow persistence (example: record in my app's domain that the tenant created a 'validate department' workflow, that we handed it off to the elsa workflow persistence service with some specific name etc and that we consider the workflow to be of a certain type "department validation")

3) These validation workflows will always be called internally by my applications services, I don't need to trigger them via an http end point or api endpoint, I will use the c# StartWorkFlow method, so these workflows don't need to start with a 'Signal Received' or 'Http Endpoint' activities.  I am seeing that if I persist a workflow in the workflow definition store (we are using SQLServer) that these workflows (that don't have blocking activities as their first step) get run at server startup time.  Am I correct that Elsa is checking the definition store and running all workflows stored there?  If so how do I get the elsa service not to do that?  If it is not doing that then any clues why my workflow definitions are getting run at startup? (again my code will manually run them when needed for business logic execution and they should not run at startup).

Sorry this is so long but I wanted to explain the use case so it makes sense.

Thanks,
Steve

**Comment by stevengwebsolutions** [comment]
Posting some answers to my own questions (in reverse order):

- just to restate my goal is to create business logic workflows that I will invoke directly from my applications code (no triggering workflows from Http endpoints in this scenario).  These types of workflows will not have blocking activities in them therefor I don't want them published in the standard Elda Workflow Definition store (as I don't want them randomly running when Elsa starts up). Lastly I want to extend the Workflow Definition schema so that I can find workflows of a given type (ie. FindWorkflow of type 'ValidateShoppingCartData' that has 'CustomerType=vip') 

3) I have implemented my own WorkflowDefinition sql table that is based on the Elsa workflow definition schema but also includes my extensions (ie. WorkflowType, CustomerType etc).  I have implemented my own Application Service library that can be used to get workflows based on Id's but also based in the schema extensions shown above.  Once I get the workflow from my definition table I can pass it to the core Elsa service to run it.  When users want to create workflows of a specific type I create an empty workflow of that type (this way I know it's Id) and put it in my table and then point the workflow editor at it (see below) so the user can customize it.

2) I have implemented my own 'WorkflowEditorApiController' which has the needed end-points defined in it that the Elsa Workflow Editor calls during edit sessions.  For example:

```
[HttpGet("workfloweditorapi/v1/workflow-definitions/{workflowDefinitionId}/{versionOptions}")]
        public async Task<IActionResult> GetByIdAsync(string workflowDefinitionId, VersionOptions versionOptions, CancellationToken cancellationToken = default)
{ .. code left out ..}

[HttpPost("workfloweditorapi/v1/workflow-definitions")]
        public async Task<ActionResult<WorkflowDefinition>> SaveAsync([FromBody] SaveWorkflowDefinitionRequest request, [FromRoute] ApiVersion apiVersion, CancellationToken cancellationToken)
{.. code left out ..}
```
I point the editor at my Api endpoints by changing the serverUrl that you pass to the designer on the razor page:

`var serverUrl = _configuration["Elsa:Server:BaseUrl"] + "/workfloweditorapi";`

('/workfloweditorapi' added to the api base url)
 
```
<elsa-studio-root server-url="@serverUrl" monaco-lib-path="_content/Elsa.Designer.Components.Web/monaco-editor/min" base-path="/appWorkflow/Editor">
        <elsa-workflow-definition-editor-screen server-url="@serverUrl" workflow-definition-id="@definitionId" monaco-lib-path="_content/Elsa.Designer.Components.Web/monaco-editor/min" />
    </elsa-studio-root>
```

The editor calls my api and then my api gets the requested workflow-definition from my workflow definition table.

1) The above code snippet shows how I am imbedding just the editor onto a razor page.  I pass in the workflow-definition-id as a pageURL parameter so we know which workflow to edit.  I have a close button on the page that takes the user back from where they came (in my case another page that has a grid of workflows of type 'ValidateShoppingCartData' that can each be customized).  My workflow definition table also supports muti-tenancy so each tenant of the app has their own set of workflows.

**Reply by sfmskywalker** [reply]
Hi @stevengwebsolutions ,

Any solution that works for you is great, but let me revisit each question. Perhaps you find some useful additional information.

> 1. Is it possible to use the workflow designer outside of the full workflow dashboard (on my 'manage department validation workflow' page I want to let the tenant launch the designer to specifically create a workflow of that type so my code can handle the tracking and persistence of that workflow type for the tenant.

Yes, for this you would use the following markup:

```html
<elsa-studio-root server-url="//your-elsa-host-address">
   <elsa-workflow-definition-editor-screen workflow-definition-id="an ID or empty for new ones" ></elsa-workflow-definition-editor-screen>
</elsa-studio-root>
```

> 2. Is it possible to point the designer at a faux server so that again I can encapsulate the designers interactions and add an additional management layer to workflow persistence (example: record in my app's domain that the tenant created a 'validate department' workflow, that we handed it off to the elsa workflow persistence service with some specific name etc and that we consider the workflow to be of a certain type "department validation")

Yes, you can use any server as long as it exposes a API endpoints that the designer expects to communicate with. I think that the easiest approach might be to use the Elsa API endpoints as-is and handle domain events published by various Elsa services for your application to handle. Elsa uses MediatR to publish various notifications. You could take this to your advantage to handle the `WorkflowDefinitionPublished` event, in response to which you would create some record with metadata into your own table (of the same database or a different one, doesn't matter).

> 3. These validation workflows will always be called internally by my applications services, I don't need to trigger them via an http end point or api endpoint, I will use the c# StartWorkFlow method, so these workflows don't need to start with a 'Signal Received' or 'Http Endpoint' activities. I am seeing that if I persist a workflow in the workflow definition store (we are using SQLServer) that these workflows (that don't have blocking activities as their first step) get run at server startup time. Am I correct that Elsa is checking the definition store and running all workflows stored there? If so how do I get the elsa service not to do that? If it is not doing that then any clues why my workflow definitions are getting run at startup? (again my code will manually run them when needed for business logic execution and they should not run at startup).

The reason your workflows start automatically is because of the `StartWorkflow` method - what this method does is register an `IHostedService` implementation that runs the specified workflow at startup. To prevent the server from running workflows automatically, simply don't use this method.

Instead, to programmatically execute workflows, use the `IStartsWorkflow` service and the `IWorkflowRegistry` to find the workflow to execute. For instance, if you have a workflow definition named "ValidateShoppingCartData" for tenant 99, you can run this workflow like this:

```csharp
var workflow = await _workflowRegistry.FindAsync(x => x.Name == "ValidateShoppingCartData" && x.TenantId == "99" && x.IsPublished);

var result = await _workflowStarter.StartWorkflowAsync(workflow);
```
